extension/src/background/index.ts

```diff
@@ -135,57 +128,60 @@ window.getSwapToSign = async () => {
     return swapToSign;
 };
 // @ts-ignore
-window.signAndSendSwap = async (txHex: string, tabId: number) => {
-    let payload;
-    let err;
-
-    try {
-        payload = await signAndSendSwap(walletName, txHex);
-    } catch (e) {
-        error(e);
-        err = e;
+window.signAndSendSwap = (txHex: string) => {
+    if (!resolveSwapSignRequest || !rejectSwapSignRequest) {
+        throw new Error("No pending promise functions for swap sign request");
     }
 
-    browser.tabs.sendMessage(tabId, { direction: Direction.ToPage, kind: MessageKind.SwapTxid, payload, error: err });
-    swapToSign = undefined;
-    updateBadge();
+    signAndSendSwap(walletName, txHex).then(resolveSwapSignRequest).catch(rejectSwapSignRequest).then(() => {
+        resolveSwapSignRequest = null;
+        rejectSwapSignRequest = null;
+        swapToSign = null;
+        updateBadge();
+    });
 };
 // @ts-ignore
-window.rejectSwap = (tabId: number) => {
-    browser.tabs.sendMessage(tabId, { direction: Direction.ToPage, kind: MessageKind.SwapRejected });
-    swapToSign = undefined;
+window.rejectSwap = () => {
+    if (!resolveSwapSignRequest || !rejectSwapSignRequest) {
+        throw new Error("No pending promise functions for swap sign request");
+    }
+
+    rejectSwapSignRequest("User declined signing request");
+    rejectSwapSignRequest = null;
+    swapToSign = null;
+
     updateBadge();
 };
 // @ts-ignore
 window.getLoanToSign = () => {
     return loanToSign;
 };
 // @ts-ignore
-window.signLoan = async (tabId: number) => {
+window.signLoan = async () => {
+    if (!resolveLoanSignRequest || !rejectLoanSignRequest) {
+        throw new Error("No pending promise functions for loan sign request");
+    }
+
     // TODO: Currently, we assume that whatever the user has verified
     // on the pop-up matches what is stored in the extension's
     // storage. It would be better to send around the swap ID to check
     // that the wallet is signing the same transaction the user has authorised
-
-    let payload;
-    let err;
-
-    try {
-        payload = await signLoan(walletName);
-    } catch (e) {
-        error(e);
-        err = e;
-    }
-
-    browser.tabs.sendMessage(tabId, { direction: Direction.ToPage, kind: MessageKind.SignedLoan, payload, error: err });
-    loanToSign = undefined;
-    updateBadge();
+    signLoan(walletName).then(resolveLoanSignRequest).catch(rejectLoanSignRequest).then(() => {
```

Comment:
If `catch` is called, we do not call `updateBadge` and the extension will always show the badge sign.

Reply:
Ah yes good _catch_! Will fix.

This state management is something I am hoping to fix by introducing a popup for signing stuff. However, as noted in https://github.com/comit-network/waves/issues/228, we can't use the wallet from two different pages at the moment because it would be loaded again due to how the `wasm-proxy` works at the moment.

Reply:
Should be fixed in https://github.com/comit-network/waves/pull/227/commits/318d8c4646c79243e2e8cbc43cd73d1d7986be14.